File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_a3/organic_2.py
# Linked List Reversal - organic submission 2
# Purely functional style, no classes wrapping the list; plain nodes + free functions.

class Node:
    def __init__(self, value, nxt=None):
        self.value = value
        self.nxt = nxt


def from_values(values):
    head = None
    for v in reversed(values):
        head = Node(v, head)
    return head


def reverse_list(head):
    previous, current = None, head
    while current:
        current.nxt, previous, current = previous, current, current.nxt
    return previous


def print_list(head):
    out = []
    while head:
        out.append(head.value)
        head = head.nxt
    print(out)


def main():
    head = from_values([5, 10, 15, 20])
    print_list(head)
    head = reverse_list(head)
    print_list(head)


if __name__ == "__main__":
    main()

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_a3/organic_2.java
// Linked List Reversal - organic submission 2
// Functional style, plain nodes + free functions.

class Node {
    int value;
    Node nxt;

    public Node(int value, Node nxt) {
        this.value = value;
        this.nxt = nxt;
    }

    public Node(int value) {
        this(value, null);
    }
}

class Solution {
    public static Node fromValues(int[] values) {
        Node head = null;
        for (int i = values.length - 1; i >= 0; i--) {
            head = new Node(values[i], head);
        }
        return head;
    }

    public static Node reverse(Node head) {
        Node previous = null;
        Node current = head;
        while (current != null) {
            Node following = current.nxt;
            current.nxt = previous;
            previous = current;
            current = following;
        }
        return previous;
    }

    public static void show(Node head) {
        Node cur = head;
        StringBuilder sb = new StringBuilder();
        while (cur != null) {
            sb.append(cur.value);
            if (cur.nxt != null) sb.append(" -> ");
            cur = cur.nxt;
        }
        System.out.println(sb.toString());
    }

    public static void main(String[] args) {
        Node head = fromValues(new int[]{10, 20, 30, 40});
        show(head);
        head = reverse(head);
        show(head);
    }
}
